cri/v1alpha2/cri_network.go

```diff
@@ -0,0 +1,171 @@
+package v1alpha2
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/alibaba/pouch/cri/config"
+
+	"github.com/cri-o/ocicni/pkg/ocicni"
+	"github.com/sirupsen/logrus"
+	runtime "k8s.io/kubernetes/pkg/kubelet/apis/cri/runtime/v1alpha2"
+)
+
+// CniMgr as an interface defines all operations against CNI.
+type CniMgr interface {
+	// Name returns the plugin's name. This will be used when searching
+	// for a plugin by name, e.g.
+	Name() string
+
+	// SetUpPodNetwork is the method called after the sandbox container of the
+	// pod has been created but before the other containers of the pod
+	// are launched.
+	SetUpPodNetwork(podNetwork *ocicni.PodNetwork) error
+
+	// TearDownPodNetwork is the method called before a pod's sandbox container will be deleted.
+	TearDownPodNetwork(podNetwork *ocicni.PodNetwork) error
+
+	// GetPodNetworkStatus is the method called to obtain the ipv4 or ipv6 addresses of the pod sandbox.
+	GetPodNetworkStatus(netnsPath string) (string, error)
+
+	// Status returns error if the network plugin is in error state.
+	Status() error
+}
+
+// CniManager is an implementation of interface CniMgr.
+type CniManager struct {
+	// plugin is used to setup and teardown network when run/stop pod sandbox.
+	plugin ocicni.CNIPlugin
+}
+
+// NewCniManager initializes a brand new cni manager.
+// If initialize failed, return NoopCniManager, we should not make pouchd creashed
+// because of the failure of cni manager.
+func NewCniManager(cfg *config.Config) CniMgr {
+	networkPluginBinDir := cfg.NetworkPluginBinDir
+	networkPluginConfDir := cfg.NetworkPluginConfDir
+
+	// Create CNI configuration directory if it doesn't exist to avoid breaking.
+	_, err := os.Stat(networkPluginConfDir)
+	if err != nil && os.IsNotExist(err) {
+		err = os.MkdirAll(networkPluginConfDir, 0666)
+		if err != nil {
+			logrus.Errorf("failed to create configuration directory for CNI: %v", err)
+			return &NoopCniManager{}
+		}
+	}
+
+	plugin, err := ocicni.InitCNI(networkPluginConfDir, networkPluginBinDir)
+	if err != nil {
+		logrus.Errorf("failed to initialize cni manager: %v", err)
+		return &NoopCniManager{}
+	}
+
+	return &CniManager{
+		plugin: plugin,
+	}
+}
+
+// Name returns the plugin's name. This will be used when searching
+// for a plugin by name, e.g.
+func (c *CniManager) Name() string {
+	return c.plugin.Name()
+}
+
+// SetUpPodNetwork is the method called after the sandbox container of the
+// pod has been created but before the other containers of the pod
+// are launched.
+func (c *CniManager) SetUpPodNetwork(podNetwork *ocicni.PodNetwork) error {
+	_, err := c.plugin.SetUpPod(*podNetwork)
+	if err != nil {
+		return fmt.Errorf("failed to setup network for sandbox %q: %v", podNetwork.ID, err)
+	}
+
+	defer func() {
+		if err != nil {
```

Comment:
I don't think the `defer` function work here, because it has finished if the err is not null.